contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,959 @@
+
+    error NotEnoughWethForSwap(uint256 wethBalance, uint256 requiredWeth); // 0x989e5ca8
+    error NotEnoughWethLiquidity(uint256 wethBalance, uint256 requiredWeth); // 0xa6737d87
+    error PoolRebalanceOutOfBounds(
+        uint256 currentPoolWethShare,
+        uint256 allowedWethShareStart,
+        uint256 allowedWethShareEnd
+    ); // 0x3681e8e0
+    error OutsideExpectedTickRange(int24 currentTick); // 0x46a58db6
+
+    event PoolRebalanced(uint256 currentPoolWethShare);
+
+    event PoolWethShareIntervalUpdated(
+        uint256 allowedWethShareStart,
+        uint256 allowedWethShareEnd
+    );
+
+    event PrincipalPositionAfterSwap(
+        uint256 wethPositionBalance,
+        uint256 oethbPositionBalance
+    );
+    event LiquidityRemoved(
+        uint256 withdrawLiquidityShare,
+        uint256 removedWETHAmount,
+        uint256 removedOETHbAmount,
+        uint256 wethAmountCollected,
+        uint256 oethbAmountCollected,
+        uint256 underlyingAssets
+    );
+
+    event LiquidityAdded(
+        uint256 wethAmountDesired,
+        uint256 oethbAmountDesired,
+        uint256 wethAmountSupplied,
+        uint256 oethbAmountSupplied,
+        uint256 tokenId,
+        uint256 underlyingAssets
+    );
+
+    event UnderlyingAssetsUpdated(uint256 underlyingAssets);
+
+    /**
+     * @dev Verifies that the caller is the Governor, or Strategist.
+     */
+    modifier onlyGovernorOrStrategist() {
+        require(
+            msg.sender == governor() ||
+                msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Not the Governor or Strategist"
+        );
+        _;
+    }
+
+    /**
+     * @dev Un-stakes the token from the gauge for the execution duration of
+     * the function and after that re-stakes it back in.
+     *
+     * It is important that the token is unstaked and owned by the strategy contract
+     * during any liquidity altering operations and that it is re-staked back into the
+     * gauge after liquidity changes. If the token fails to re-stake back to the
+     * gauge it is not earning incentives.
+     */
+    modifier gaugeUnstakeAndRestake() {
+        if (tokenId != 0 && lpTokenDepositedToGauge) {
+            clGauge.withdraw(tokenId);
+            lpTokenDepositedToGauge = false;
+        }
+        _;
+        if (tokenId != 0 && !lpTokenDepositedToGauge) {
+            (
+                uint256 _wethPositionBalance,
+                uint256 _oethbPositionBalance
+            ) = getPositionPrincipal();
+
+            /**
+             * It can happen that a withdrawal (or a full withdrawal) transactions would
+             * remove all of the liquidity from the token with a NFT token still existing.
+             * In that case the token can not be staked into the gauge, as some liquidity
+             * needs to be added to it first.
+             */
+            if (_wethPositionBalance + _oethbPositionBalance > 0) {
```

Comment:
nit: any reason for not checking `_getLiquidity() > 0` check instead? Since we are least concerned about gas, either is fine. But just curious to know the idea behind using `getPositionPrinicipal` over just checking if liquidity is non-zero

Reply:
Both work equally well. Changing to `getLiquidity` as suggested, since it seems cleaner: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/053c201868ff74f4e579bcaf5317bf184e83af13